core/templates/container_converters.h

```diff
@@ -0,0 +1,103 @@
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef CONTAINER_CONVERTERS_H
+#define CONTAINER_CONVERTERS_H
+
+#include "core/variant/typed_array.h"
+
+namespace converters {
+
+template <typename T>
+TypedArray<T> vector_to_typed_array(const Vector<T> &p_vector) {
+	TypedArray<T> ret;
+	ret.resize(p_vector.size());
+
+	Array::Iterator itr = ret.begin();
+	for (const T &E : p_vector) {
+		*itr = E;
+		++itr;
+	}
+	return ret;
+}
+
+template <typename T>
+List<T> vector_to_list(const Vector<T> &p_vector) {
+	List<T> ret;
+
+	for (const T &E : p_vector) {
+		ret.push_back(E);
+	}
+	return ret;
+}
+
+template <typename R, typename T>
+Vector<R> vector_change_type(const Vector<T> &p_vector) {
+	Vector<R> ret;
+	ret.resize(p_vector.size());
+
+	R *write = ret.ptrw();
+	for (const T &E : p_vector) {
+		*write = E;
+		write++;
+	}
+	return ret;
+}
+
+template <typename T>
+Vector<T> list_to_vector(const List<T> &p_list) {
+	Vector<T> ret;
+	ret.resize(p_list.size());
```

Comment:
The resize _may_ be a bad idea if the type is not trivially destructible, because it will initialize the entries and then destruct them in the subsequent loop. Since the lists have no concept of `reserve()` (yet?) the only way around this that I can think of is using an 
```c++
if constexpr (std::is_trivially_destructible<T>) { ... } else { ... }
```

Not sure if it's worth the fuss but it may become a problem performance wise if the function ends up being used uncarefully.

Reply:
`resize()` does not initialize data, `resize_zeroed()` does.

Reply:
That is not true for `CowData`: 

https://github.com/godotengine/godot/blob/19e003bc08ac2fb14fc2434bde9530387afd994c/core/templates/cowdata.h#L369-L375

Reply:
Maybe we should have a new method that resizes *never* initializing?
What's the alternative? Appending each element? That's probably even worse.

Reply:
CowData still has a capacity > size... Perhaps it would be possible to patch in a `reserve()` function. Otherwise, just calling append wouldn't be worse than the current state of things anyway.

Reply:
> just calling append wouldn't be worse than the current state of things anyway.

But it wouldn't be better either, no?
Improving Vector is out of scope of this PR and can be done separately. Note that we already use `resize()` in various places. Since the new methods are supposed to be used universally, it will be easier to improve it all at once.

Reply:
Yeah i agree it's out of scope to attempt to fix it now. Since most Godot types are CoW with a null state it probably won't be too bad to just use resize regardless. Perhaps add a comment about it for future improvements and call it a day :)